nexus/db-queries/src/db/datastore/region_replacement.rs

```diff
@@ -732,9 +732,18 @@ impl DataStore {
     }
 
     /// Nexus has been notified by an Upstairs (or has otherwised determined)
-    /// that a region replacement is done, so update the record. This may arrive
-    /// in the middle of a drive saga invocation, so do not filter on state or
-    /// operating saga id!
+    /// that a region replacement is done, so update the record. Filter on the
+    /// following:
+    ///
+    /// - operating saga id being None, as this happens outside of a saga and
+    ///   should only transition the record if there isn't currently a lock.
```

Comment:
What happens if there _is_ currently a saga holding the lock and we receive the notification from an upstairs that results in this? Do we just eat it? Or is there something else that guarantees that it's eventually observed?

It's entirely possible this is documented somewhere else, I haven't reviewed most of the rest of the code involved here...

Reply:
All good, thanks for having a look :)

The internal endpoint that receives these notifications is `cpapi_upstairs_repair_finish`, which calls `Nexus::upstairs_repair_finish`. That function only records the notification, it doesn't take any action directly.

It's the region replacement drive background task that will scan for these finish notifications and if one is found will mark the corresponding repair as done. If that fails (like it should after this PR), then a future background task invocation should eventually mark it correctly.